draft-ietf-mls-protocol.md

```diff
@@ -1810,11 +1906,68 @@ followed by reconstructing the Add and Commit messages and verifying that the
 resulting transcript hashes and epoch secret match those found in the Welcome
 message.
 
+## Group Creation from existing PSKs
+
+Group creation may be tied to an already existing group structure, consisting of
+re-initialization of an existing group, or branching of a sub-group.
+
+Re-initialization of an existing group may be used, for example, to re-start the
+group based on the current group state but under a different ciphersuite.
+
+Branching may be used to bootstrap a new group consisting of a subset of
+current group members, based on the current group state.
+
+### Re-Initialization {#re-initialization}
+
+If the group is created as the follow-up to a `Re-Init` proposal, the creator
+MUST include a `PSKId` in the Welcome message of the new group using the
+`psktype` `re-initialization`, as well as a valid epoch number within the list
+of maintained `recovery_secret` secrets and the group ID of the group they want
+to re-initialize. They also must include a nonce to avoid key re-use.
+
+~~~~~
+struct {
+  PSKType psktype;
+  select (psktype) {
+
+    case external:
+      opaque psk_id<0..255>;
```

Comment:
Are "external" PSKs fixed? Would it make sense for them to also include an epoch, to allow applications to roll them forward for some FS?

Reply:
External PSKs are required to be of an "equivalent security level" (although that terminology was added to the draft a long time ago and not yet defined). As such, any FS guarantees or other requirements must be met before being injected. It does not make sense to ratchet external PSK forward per MLS spec, as then we would be trying to correlate the time of existence of the PSK to MLS (i.e. parallel ratcheting); in that case the PSK is no longer external to the protocol but somehow correlated to it. The nonce prevents reuse, but any additional guarantees should be met by the application, not the protocol (going back to meeting requirements before being injected).

Reply:
I’m not sure I agree. You can certainly do the ratcheting outside of the protocol, but you can also do it inside. For example, send an epoch and psk_id, and hash forward the external PSK corresponding to the ID based on the epoch. Some applications might not be able to augment their external PSK provision machinery to ratchet, so handling it in the protocol can be appealing. Food for thought.

Reply:
So the proposal would be to extend MLS to allow ratcheting of "external" PSKs that continuously get injected into the protocol?

It's interesting, but it seems to me like we'd be creating a degree of redundancy here. MLS doesn't have very good FS capabilities (although they can be improved by something like RTreeKEM), so it's tempting to create a secondary mechanism, i.e. ratcheting a PSK along the epoch-chain. However, then the FS guarantees hinge on the FS properties of whatever means was used to distribute the PSK. So what is a scenario where the means of PSK distribution gives you better FS than MLS? If you have this really good out-of-band distribution channel for PSKs, do you really still need MLS?

I don't think it's a bad idea, but I'm not sure how often it will be used and if it warrants the added complexity in the protocol. But that's maybe an issue we should discuss on the list or in the next virtual interim? 

In any case, I think it's a good idea that's worth thinking about and that might match the use case of a number of people, but it's relatively orthogonal enough to our PR. Maybe it would be worth a new github issue?